### DISCUSSION TITLE
Removing StateTransition checks from kill/delete client functions?

### DISCUSSION THREAD

**Discussion by sfayer** [DIRACGrid/DIRAC#6663] [General]
Hi,

The stable `DIRAC/src/DIRAC/Interfaces/API/Dirac.py` killJob and deleteJob functions do a transition check to filter the list of jobs for ones that can be killed/deleted respectively. There are a couple of side-effects of doing this:
- You can't use killJob on a job that's in the waiting state: The server side does a check and will delete the job instead, but the filter in the client stops this from happening.
- If all jobs are removed from the list by the filter, the client will post an empty list, resulting in the confusing error `Invalid job specification: []`
- The documentation for deleteJob claims "If running, these jobs will be first killed", but this is no-longer true: The JobManager will do this but the filter removes running jobs from the list so it doesn't happen.

I propose removing the filter from these two functions: There is enough filtering done on the server side to handle all of the possible cases in a sensible way, the extra client-side filtering doesn't seem to add very much (from my perspective), what do people think?

Regards,
Simon

**Comment by fstagni** [comment]
This check was introduced in https://github.com/DIRACGrid/DIRAC/pull/5752/files. IIRC (I might be wrong), the reason for introducing this check is that calling commands like `dirac-wms-job-kill` on a non-kill-able job would have returned an `S_OK` result (while the status would have not changed).

**Comment by sfayer** [comment]
Understood... Perhaps we could do one of the following?

Either extend the filter in both cases so that it checks for both possible states that may be applied?
e.g. something like:
```diff
-            res = JobStatus.checkJobStateTransition(jobID, JobStatus.KILLED)
-            if res["OK"]:
-                jobIDsToKill.append(jobID)
+            can_kill = JobStatus.checkJobStateTransition(jobID, JobStatus.KILLED)["OK"]
+            can_del = JobStatus.checkJobStateTransition(jobID, JobStatus.DELETED)["OK"]
+            if can_kill or can_del:
+                jobIDsToKill.append(jobID)
```

Or remove the client filter and modify `dirac-wms-job-kill` and `...job-delete` so that they process the `result["InvalidJobIDs"]` field to tell the user if there were jobs that didn't get updated?

Regards,
Simon

**Reply by fstagni** [reply]
>  extend the filter in both cases so that it checks for both possible states that may be applied? e.g. something like:
> 
> ```diff
> -            res = JobStatus.checkJobStateTransition(jobID, JobStatus.KILLED)
> -            if res["OK"]:
> -                jobIDsToKill.append(jobID)
> +            can_kill = JobStatus.checkJobStateTransition(jobID, JobStatus.KILLED)["OK"]
> +            can_del = JobStatus.checkJobStateTransition(jobID, JobStatus.DELETED)["OK"]
> +            if can_kill or can_del:
> +                jobIDsToKill.append(jobID)
> ```> 

This, I would say.

**Reply by fstagni** [reply]
Would you take care of the PR?

**Reply by sfayer** [reply]
Sure, #6665 .
